### DISCUSSION TITLE
Convergence Issues with a Second-Order Mixed Partial Derivative in MOOSE

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#28265] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I am currently working on solving an equation with a second-order mixed partial derivative: 
![20240730225311](https://github.com/user-attachments/assets/4fbefe5a-cc66-42ae-b17c-94a4eab188ad)
The weak form of this equation is as follows: 
![20240730225327](https://github.com/user-attachments/assets/84f00e7e-60c4-4097-b72e-d2d1edfb4639)

However, I am encountering convergence issues during computation and can only use JFNK, not PJFNK. I have tried different approaches, including directly computing the residual for the term with the mixed partial derivative using `_var.secondSln()`, as well as integrating by parts to reduce the mixed partial derivative term to two first-order partial derivative terms. 
![20240730230134](https://github.com/user-attachments/assets/154da7f9-ef51-4b3b-a1bd-efac3442fb07)
Despite these attempts, convergence remains unattainable.

In my tests, if I simplify the problem to Mat(x,y)∂F(x,y)/∂y=Q, I can achieve correct convergence by directly computing the residual term for the left-hand side of the equation using `_grad_u` in the Kernel. However, this approach also has issues since it still requires JFNK instead of PJFNK. When attempting to solve using PJFNK, I receive the following errors:
```
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT 
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
```

Does anyone know how to address this issue?

**Comment by GiudGiud** [comment]
Hello

Are you using automatic differentiation?
Maybe the Jacobian has a large mistake.
There is some advice on convergence issues here:
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

**Reply by hityyds** [reply]
No, I provide the computeQpJacobian() function manually.

**Reply by GiudGiud** [reply]
then there may be a mistake, or a missing off-diagonal term.
The Jacobian tester or the petsc option to test Jacobians are a good idea to see this

**Reply by hityyds** [reply]
I thought there was no need to use the computeQpOffDiagJacobian() function because there is only one variable, F(x,y).

**Reply by GiudGiud** [reply]
if there is only one variable and no element-to-neighbor coupling you should be fine

**Reply by hityyds** [reply]
What do you mean by element-to-neighbor? Do you mean partial derivatives?

**Reply by GiudGiud** [reply]
I mean techniques like mortar coupling or specific treatment at interface kernels

**Reply by hityyds** [reply]
I only used Kernel, BC, and DGKernel, without any complex techniques. This equation, aside from having a mixed partial derivative, doesn't have any other complexities. I'm also not sure if such an equation requires any special stabilization techniques for computation.

**Reply by GiudGiud** [reply]
I can see DGKernels have off diagonal coupling between the element and the side DoFs. 

You should look at example nonAD DG kernels to see how this is handled
